File: backend/relayer_server.ts
/**
 * relayer_server.ts — Express Relayer Service for Valentin
 *
 * Accepts optimized portfolio weights from Valentin's Python/Rust
 * backend via REST, builds PTBs, signs with AgentCap key, and
 * submits to Sui testnet.
 *
 * Endpoints:
 *   POST /api/trade       — Execute a quantum-optimized trade
 *   POST /api/audit       — Log a quantum proof hash on-chain
 *   POST /api/pause       — Emergency kill-switch (admin only)
 *   GET  /api/status      — Portfolio status + gas balance
 *   GET  /api/health      — Health check
 *
 * Start:
 *   npx ts-node relayer_server.ts
 *   # or: npm run relayer:server
 */

// eslint-disable-next-line @typescript-eslint/no-require-imports
const express = require('express');
import { Request, Response } from 'express';
import {
  SuiClient,
  getFullnodeUrl,
} from '@mysten/sui/client';
import { Transaction } from '@mysten/sui/transactions';
import { Ed25519Keypair } from '@mysten/sui/keypairs/ed25519';
import { fromBase64 } from '@mysten/sui/utils';
import * as crypto from 'crypto';
import 'dotenv/config';
import { parseAbortError, errorResponseBody, logError } from './error_map';

// ═══════════════════════════════════════════════════════════
//  CONFIG
// ═══════════════════════════════════════════════════════════

const PORT         = parseInt(process.env.RELAYER_PORT ?? '3001', 10);
const NETWORK      = (process.env.SUI_NETWORK as 'devnet' | 'testnet' | 'mainnet') ?? 'devnet';
const RPC_URL      = process.env.SUI_RPC_URL ?? getFullnodeUrl(NETWORK);

const PACKAGE_ID   = process.env.PACKAGE_ID!;
const PORTFOLIO_ID = process.env.PORTFOLIO_ID ?? process.env.PORTFOLIO_OBJECT_ID!;
const AGENT_CAP_ID = process.env.AGENT_CAP_ID!;
const ADMIN_CAP_ID = process.env.ADMIN_CAP_ID!;
const SUI_CLOCK    = '0x6';

// ═══════════════════════════════════════════════════════════
//  KEYPAIR
// ═══════════════════════════════════════════════════════════

function loadKeypair(envKey: string): Ed25519Keypair {
  const key = process.env[envKey];
  if (!key) throw new Error(`${envKey} not set in .env`);

  if (key.startsWith('suiprivkey')) {
    return Ed25519Keypair.fromSecretKey(key);
  }
  const bytes = key.startsWith('0x')
    ? Uint8Array.from(Buffer.from(key.slice(2), 'hex'))
    : fromBase64(key);
  return Ed25519Keypair.fromSecretKey(bytes);
}

const agentKeypair = loadKeypair('AGENT_PRIVATE_KEY');
const client       = new SuiClient({ url: RPC_URL });

// Try to load admin keypair (only Korbinian has this)
let adminKeypair: Ed25519Keypair | null = null;
try {
  adminKeypair = loadKeypair('ADMIN_PRIVATE_KEY');
} catch {
  console.log('⚠️  ADMIN_PRIVATE_KEY not set — pause endpoint disabled');
}

// ═══════════════════════════════════════════════════════════
//  EXPRESS APP
// ═══════════════════════════════════════════════════════════

const app = express();
app.use(express.json());

// ── Health check ────────────────────────────────────────────

app.get('/api/health', (_req: Request, res: Response) => {
  res.json({
    status: 'ok',
    network: NETWORK,
    rpc: RPC_URL,
    package: PACKAGE_ID,
    agent: agentKeypair.getPublicKey().toSuiAddress(),
    timestamp: new Date().toISOString(),
  });
});

// ── Portfolio status ────────────────────────────────────────

app.get('/api/status', async (_req: Request, res: Response) => {
  try {
    const portfolio = await client.getObject({
      id: PORTFOLIO_ID,
      options: { showContent: true },
    });

    const fields = (portfolio.data?.content as any)?.fields ?? {};

    // Agent gas balance
    const agentAddr = agentKeypair.getPublicKey().toSuiAddress();
    const coins = await client.getCoins({ owner: agentAddr, coinType: '0x2::sui::SUI' });
    const gasBalance = coins.data.reduce(
      (acc: bigint, c: any) => acc + BigInt(c.balance), BigInt(0),
    );

    res.json({
      portfolio: {
        id: PORTFOLIO_ID,
        balance: fields.balance,
        peak_balance: fields.peak_balance,
        trade_count: fields.trade_count,
        paused: fields.paused,
        max_drawdown_bps: fields.max_drawdown_bps,
        daily_volume_limit: fields.daily_volume_limit,
        cooldown_ms: fields.cooldown_ms,
        total_traded_today: fields.total_traded_today,
      },
      agent: {
        address: agentAddr,
        gas_balance_mist: gasBalance.toString(),
        gas_balance_sui: (Number(gasBalance) / 1e9).toFixed(4),
      },
    });
  } catch (e: any) {
    res.status(500).json({ error: e.message });
  }
});

// ── Execute trade ───────────────────────────────────────────

interface TradeRequest {
  /** Amount in MIST (1 SUI = 1_000_000_000 MIST) */
  amount: number;
  /** Minimum acceptable output — slippage protection */
  min_output: number;
  /** 0–100 quantum optimization score */
  quantum_score?: number;
  /** true if based on quantum RNG */
  is_quantum_optimized?: boolean;
  /** Optional: raw QUBO solution data for proof hashing */
  qubo_solution_data?: string;
}

app.post('/api/trade', async (req: Request, res: Response) => {
  try {
    const body = req.body as TradeRequest;

    if (!body.amount || body.amount <= 0) {
      res.status(400).json({ error: 'amount must be > 0' });
      return;
    }
    if (body.min_output === undefined || body.min_output < 0) {
      res.status(400).json({ error: 'min_output is required and must be >= 0' });
      return;
    }

    const amount            = body.amount;
    const minOutput         = body.min_output;
    const quantumScore      = body.quantum_score ?? 0;
    const isQuantumOptimized = body.is_quantum_optimized ?? true;

    console.log(`\n📥 Trade request: ${amount} MIST, min_output=${minOutput}, q_score=${quantumScore}`);

    // ── Build PTB ──
    const tx = new Transaction();

    tx.moveCall({
      target: `${PACKAGE_ID}::portfolio::swap_and_rebalance`,
      arguments: [
        tx.object(AGENT_CAP_ID),
        tx.object(PORTFOLIO_ID),
        tx.pure.u64(amount),
        tx.pure.u64(minOutput),
        tx.pure.bool(isQuantumOptimized),
        tx.pure.u64(quantumScore),
        tx.object(SUI_CLOCK),
      ],
    });

    // If QUBO solution data provided, also log a quantum audit receipt
    if (body.qubo_solution_data) {
      const proofHash = crypto
        .createHash('sha256')
        .update(body.qubo_solution_data)
        .digest();

      tx.moveCall({
        target: `${PACKAGE_ID}::audit_trail::log_execution`,
        arguments: [
          tx.object(AGENT_CAP_ID),
          tx.pure('vector<u8>', Array.from(proofHash)),
          tx.pure.u64(amount),
          tx.pure.u64(quantumScore),
          tx.object(SUI_CLOCK),
        ],
      });
    }

    // ── Sign & submit ──
    const result = await client.signAndExecuteTransaction({
      signer: agentKeypair,
      transaction: tx,
      options: { showEffects: true, showEvents: true },
    });

    const status = result.effects?.status?.status;
    const events = result.events?.map((ev) => ({
      type: ev.type,
      data: ev.parsedJson,
    })) ?? [];

    console.log(`✅ TX ${result.digest} — status: ${status}`);

    res.json({
      success: status === 'success',
      digest: result.digest,
      status,
      events,
      error: status !== 'success'
        ? result.effects?.status?.error
        : undefined,
    });
  } catch (e: any) {
    logError('Trade', e);
    const parsed = parseAbortError(e);
    const status = parsed.isMoveAbort ? 422 : 500;
    res.status(status).json(errorResponseBody(e));
  }
});

// ── Log quantum audit proof (standalone) ────────────────────

interface AuditRequest {
  proof_data: string;      // raw data — will be SHA-256 hashed
  amount: number;
  quantum_score?: number;
}

app.post('/api/audit', async (req: Request, res: Response) => {
  try {
    const body = req.body as AuditRequest;

    if (!body.proof_data) {
      res.status(400).json({ error: 'proof_data is required' });
      return;
    }

    const proofHash = crypto
      .createHash('sha256')
      .update(body.proof_data)
      .digest();

    const tx = new Transaction();

    tx.moveCall({
      target: `${PACKAGE_ID}::audit_trail::log_execution`,
      arguments: [
        tx.object(AGENT_CAP_ID),
        tx.pure('vector<u8>', Array.from(proofHash)),
        tx.pure.u64(body.amount ?? 0),
        tx.pure.u64(body.quantum_score ?? 0),
        tx.object(SUI_CLOCK),
      ],
    });

    const result = await client.signAndExecuteTransaction({
      signer: agentKeypair,
      transaction: tx,
      options: { showEffects: true, showEvents: true },
    });

    console.log(`📝 Audit TX: ${result.digest}`);

    res.json({
      success: result.effects?.status?.status === 'success',
      digest: result.digest,
      proof_hash_hex: proofHash.toString('hex'),
    });
  } catch (e: any) {
    logError('Audit', e);
    res.status(500).json(errorResponseBody(e));
  }
});

// ── Emergency pause (admin only) ────────────────────────────

app.post('/api/pause', async (req: Request, res: Response) => {
  if (!adminKeypair) {
    res.status(403).json({ error: 'Admin key not configured on this relayer' });
    return;
  }

  try {
    const paused = req.body.paused ?? true;

    const tx = new Transaction();
    tx.moveCall({
      target: `${PACKAGE_ID}::portfolio::set_paused`,
      arguments: [
        tx.object(ADMIN_CAP_ID),
        tx.object(PORTFOLIO_ID),
        tx.pure.bool(paused),
      ],
    });

    const result = await client.signAndExecuteTransaction({
      signer: adminKeypair,
      transaction: tx,
      options: { showEffects: true, showEvents: true },
    });

    console.log(`${paused ? '🛑 PAUSED' : '▶️  RESUMED'} TX: ${result.digest}`);

    res.json({
      success: result.effects?.status?.status === 'success',
      digest: result.digest,
      paused,
    });
  } catch (e: any) {
    logError('Pause', e);
    res.status(500).json(errorResponseBody(e));
  }
});

// ═══════════════════════════════════════════════════════════
//  START
// ═══════════════════════════════════════════════════════════

app.listen(PORT, () => {
  console.log(`\n🚀 Relayer server running on http://localhost:${PORT}`);
  console.log(`   Network:   ${NETWORK}`);
  console.log(`   Package:   ${PACKAGE_ID}`);
  console.log(`   Portfolio:  ${PORTFOLIO_ID}`);
  console.log(`   Agent:      ${agentKeypair.getPublicKey().toSuiAddress()}`);
  console.log(`\n   Endpoints:`);
  console.log(`     POST /api/trade   — Execute quantum-optimized trade`);
  console.log(`     POST /api/audit   — Log quantum proof on-chain`);
  console.log(`     POST /api/pause   — Emergency kill-switch`);
  console.log(`     GET  /api/status  — Portfolio & gas status`);
  console.log(`     GET  /api/health  — Health check\n`);
});

File: backend/blockchain/relayer_server.py
#!/usr/bin/env python3
"""
relayer_server.py — FastAPI Relayer Service

Accepts optimized portfolio weights from the Python AI pipeline via REST,
builds PTBs via sui CLI, and submits to Sui testnet.

Endpoints:
    POST /api/trade              — Execute a quantum-optimized trade
    POST /api/atomic-rebalance   — Multi-swap atomic rebalance
    POST /api/oracle-swap        — Oracle-validated swap
    POST /api/audit              — Log a quantum proof hash on-chain
    POST /api/pause              — Emergency kill-switch (admin only)
    GET  /api/status             — Portfolio status + gas balance
    GET  /api/health             — Health check

Start:
    uvicorn relayer_server:app --host 0.0.0.0 --port 3001 --reload
"""

import hashlib
import json
import logging
import os
import subprocess
import time
from typing import Any, Dict, List, Optional

import httpx
from dotenv import load_dotenv
from fastapi import FastAPI, HTTPException, Request
from fastapi.middleware.cors import CORSMiddleware
from pydantic import BaseModel

from core.error_map import error_response_body, log_error, parse_abort_error
from blockchain.ptb_builder import (
    TxResult,
    _run_sui_cmd,
    build_atomic_rebalance,
    build_execute_rebalance,
    build_log_execution,
    build_oracle_validated_swap,
    build_set_paused,
    build_swap_and_rebalance,
)

load_dotenv()

logger = logging.getLogger(__name__)
logging.basicConfig(level=logging.INFO)

# ═══════════════════════════════════════════════════════════
#  CONFIG
# ═══════════════════════════════════════════════════════════

PORT = int(os.getenv("RELAYER_PORT", "3001"))
NETWORK = os.getenv("SUI_NETWORK", "devnet")
RPC_URL = os.getenv("SUI_RPC_URL", f"https://fullnode.{NETWORK}.sui.io:443")

PACKAGE_ID = os.getenv("PACKAGE_ID", "")
PORTFOLIO_ID = os.getenv("PORTFOLIO_ID", os.getenv("PORTFOLIO_OBJECT_ID", ""))
AGENT_CAP_ID = os.getenv("AGENT_CAP_ID", "")
ADMIN_CAP_ID = os.getenv("ADMIN_CAP_ID", "")
ORACLE_CONFIG_ID = os.getenv("ORACLE_CONFIG_ID", "")

# ═══════════════════════════════════════════════════════════
#  PROTOCOL WHITELIST (backend-side enforcement)
# ═══════════════════════════════════════════════════════════

# Load whitelist from env (comma-separated addresses) or default to empty (allow all)
_whitelist_env = os.getenv("PROTOCOL_WHITELIST", "")
PROTOCOL_WHITELIST: set[str] = (
    {addr.strip() for addr in _whitelist_env.split(",") if addr.strip()}
    if _whitelist_env
    else set()
)


def check_protocol_whitelist(protocol_address: Optional[str] = None):
    """
    Validate that a target protocol address is whitelisted.
    If the whitelist is empty, all protocols are allowed (demo mode).
    """
    if not PROTOCOL_WHITELIST:
        return  # empty whitelist = permissive mode
    if protocol_address and protocol_address not in PROTOCOL_WHITELIST:
        raise HTTPException(
            status_code=403,
            detail=f"Protocol {protocol_address} is not whitelisted. "
            f"Allowed: {list(PROTOCOL_WHITELIST)}",
        )

# ═══════════════════════════════════════════════════════════
#  RPC HELPER
# ═══════════════════════════════════════════════════════════

_rpc_id = 0


def _rpc_call(method: str, params: list) -> dict:
    global _rpc_id
    _rpc_id += 1
    payload = {
        "jsonrpc": "2.0",
        "id": _rpc_id,
        "method": method,
        "params": params,
    }
    with httpx.Client(timeout=15) as client:
        resp = client.post(RPC_URL, json=payload)
        resp.raise_for_status()
        data = resp.json()
    if "error" in data:
        raise RuntimeError(f"RPC error: {data['error']}")
    return data.get("result", {})


# ═══════════════════════════════════════════════════════════
#  FASTAPI APP
# ═══════════════════════════════════════════════════════════

app = FastAPI(
    title="quantum_vault Relayer",
    description="Relay service for quantum-optimized portfolio trades on Sui",
    version="2.0.0",
)

app.add_middleware(
    CORSMiddleware,
    allow_origins=["*"],
    allow_methods=["*"],
    allow_headers=["*"],
)

# ── Health check ─────────────────────────────────────────


@app.get("/api/health")
async def health():
    return {
        "status": "ok",
        "network": NETWORK,
        "rpc": RPC_URL,
        "package": PACKAGE_ID,
        "timestamp": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
    }


# ── Portfolio status ─────────────────────────────────────


@app.get("/api/status")
async def status():
    try:
        obj = _rpc_call(
            "sui_getObject",
            [
                PORTFOLIO_ID,
                {"showContent": True, "showType": True, "showOwner": True},
            ],
        )
        fields = obj.get("data", {}).get("content", {}).get("fields", {})

        return {
            "portfolio": {
                "id": PORTFOLIO_ID,
                "balance": fields.get("balance"),
                "peak_balance": fields.get("peak_balance"),
                "trade_count": fields.get("trade_count"),
                "paused": fields.get("paused"),
                "max_drawdown_bps": fields.get("max_drawdown_bps"),
                "daily_volume_limit": fields.get("daily_volume_limit"),
                "cooldown_ms": fields.get("cooldown_ms"),
                "total_traded_today": fields.get("total_traded_today"),
            },
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


# ── Execute trade ────────────────────────────────────────


class TradeRequest(BaseModel):
    amount: int
    min_output: int
    quantum_score: Optional[int] = 0
    is_quantum_optimized: Optional[bool] = True
    qubo_solution_data: Optional[str] = None
    protocol_address: Optional[str] = None  # target DEX/protocol address


@app.post("/api/trade")
async def trade(req: TradeRequest):
    if req.amount <= 0:
        raise HTTPException(status_code=400, detail="amount must be > 0")
    if req.min_output < 0:
        raise HTTPException(status_code=400, detail="min_output must be >= 0")

    # Protocol whitelist check
    check_protocol_whitelist(req.protocol_address)

    logger.info(
        f" Trade request: {req.amount} MIST, "
        f"min_output={req.min_output}, q_score={req.quantum_score}"
    )

    try:
        cmd = build_swap_and_rebalance(
            amount_mist=req.amount,
            min_output=req.min_output,
            is_quantum_optimized=req.is_quantum_optimized,
            quantum_optimization_score=req.quantum_score,
        )
        result = _run_sui_cmd(cmd)

        if not result.success:
            parsed = parse_abort_error(result.error)
            status_code = 422 if parsed.is_move_abort else 500
            raise HTTPException(status_code=status_code, detail=error_response_body(result.error))

        logger.info(f" TX {result.digest} — status: {result.status}")

        # If QUBO solution data provided, also log a quantum audit receipt
        if req.qubo_solution_data:
            proof_hash = hashlib.sha256(req.qubo_solution_data.encode()).digest()
            audit_cmd = build_log_execution(proof_hash, req.amount, req.quantum_score)
            _run_sui_cmd(audit_cmd)

        return {
            "success": True,
            "digest": result.digest,
            "status": result.status,
            "events": result.events,
        }

    except HTTPException:
        raise
    except Exception as e:
        log_error("Trade", e)
        parsed = parse_abort_error(e)
        status_code = 422 if parsed.is_move_abort else 500
        raise HTTPException(status_code=status_code, detail=error_response_body(e))


# ── Log quantum audit proof ─────────────────────────────


class AuditRequest(BaseModel):
    proof_data: str
    amount: Optional[int] = 0
    quantum_score: Optional[int] = 0


@app.post("/api/audit")
async def audit(req: AuditRequest):
    if not req.proof_data:
        raise HTTPException(status_code=400, detail="proof_data is required")

    try:
        proof_hash = hashlib.sha256(req.proof_data.encode()).digest()
        cmd = build_log_execution(proof_hash, req.amount, req.quantum_score)
        result = _run_sui_cmd(cmd)

        if not result.success:
            raise HTTPException(status_code=500, detail=error_response_body(result.error))

        logger.info(f" Audit TX: {result.digest}")

        return {
            "success": True,
            "digest": result.digest,
            "proof_hash_hex": proof_hash.hex(),
        }

    except HTTPException:
        raise
    except Exception as e:
        log_error("Audit", e)
        raise HTTPException(status_code=500, detail=error_response_body(e))


# ── Atomic rebalance ────────────────────────────────────


class AtomicRebalanceRequest(BaseModel):
    swap_amounts: List[int]
    swap_min_outputs: List[int]
    quantum_score: Optional[int] = 0
    is_quantum_optimized: Optional[bool] = True
    qubo_solution_data: Optional[str] = None


@app.post("/api/atomic-rebalance")
async def atomic_rebalance(req: AtomicRebalanceRequest):
    if not req.swap_amounts:
        raise HTTPException(status_code=400, detail="swap_amounts must be a non-empty array")
    if len(req.swap_amounts) != len(req.swap_min_outputs):
        raise HTTPException(
            status_code=400,
            detail="swap_amounts and swap_min_outputs must be same length",
        )

    logger.info(
        f" Atomic rebalance: {len(req.swap_amounts)} swaps, "
        f"total={sum(req.swap_amounts)} MIST"
    )

    try:
        cmd = build_atomic_rebalance(
            swap_amounts=req.swap_amounts,
            swap_min_outputs=req.swap_min_outputs,
            is_quantum_optimized=req.is_quantum_optimized,
            quantum_optimization_score=req.quantum_score,
        )
        result = _run_sui_cmd(cmd)

        if not result.success:
            parsed = parse_abort_error(result.error)
            status_code = 422 if parsed.is_move_abort else 500
            raise HTTPException(status_code=status_code, detail=error_response_body(result.error))

        logger.info(f" Atomic rebalance TX {result.digest} — status: {result.status}")

        # Optional audit proof
        if req.qubo_solution_data:
            proof_hash = hashlib.sha256(req.qubo_solution_data.encode()).digest()
            total_amount = sum(req.swap_amounts)
            audit_cmd = build_log_execution(proof_hash, total_amount, req.quantum_score)
            _run_sui_cmd(audit_cmd)

        return {
            "success": True,
            "digest": result.digest,
            "status": result.status,
            "num_swaps": len(req.swap_amounts),
            "events": result.events,
        }

    except HTTPException:
        raise
    except Exception as e:
        log_error("AtomicRebalance", e)
        parsed = parse_abort_error(e)
        status_code = 422 if parsed.is_move_abort else 500
        raise HTTPException(status_code=status_code, detail=error_response_body(e))


# ── Oracle-validated swap ────────────────────────────────


class OracleSwapRequest(BaseModel):
    amount: int
    min_output: int
    oracle_price_x8: int
    expected_price_x8: int
    oracle_timestamp_ms: int
    asset_symbol: str
    quantum_score: Optional[int] = 0
    is_quantum_optimized: Optional[bool] = True


@app.post("/api/oracle-swap")
async def oracle_swap(req: OracleSwapRequest):
    if req.amount <= 0:
        raise HTTPException(status_code=400, detail="amount must be > 0")
    if not req.oracle_price_x8 or not req.expected_price_x8:
        raise HTTPException(
            status_code=400,
            detail="oracle_price_x8 and expected_price_x8 required",
        )
    if not ORACLE_CONFIG_ID:
        raise HTTPException(status_code=500, detail="ORACLE_CONFIG_ID not configured")

    logger.info(
        f" Oracle swap: {req.amount} MIST, "
        f"oracle=${req.oracle_price_x8 / 1e8:.4f}, "
        f"expected=${req.expected_price_x8 / 1e8:.4f}"
    )

    try:
        cmd = build_oracle_validated_swap(
            amount_mist=req.amount,
            min_output=req.min_output,
            oracle_price_x8=req.oracle_price_x8,
            expected_price_x8=req.expected_price_x8,
            oracle_timestamp_ms=req.oracle_timestamp_ms,
            asset_symbol=req.asset_symbol,
            is_quantum_optimized=req.is_quantum_optimized,
            quantum_optimization_score=req.quantum_score,
        )
        result = _run_sui_cmd(cmd)

        if not result.success:
            parsed = parse_abort_error(result.error)
            status_code = 422 if parsed.is_move_abort else 500
            raise HTTPException(status_code=status_code, detail=error_response_body(result.error))

        logger.info(f" Oracle swap TX {result.digest} — status: {result.status}")

        return {
            "success": True,
            "digest": result.digest,
            "status": result.status,
            "events": result.events,
        }

    except HTTPException:
        raise
    except Exception as e:
        log_error("OracleSwap", e)
        parsed = parse_abort_error(e)
        status_code = 422 if parsed.is_move_abort else 500
        raise HTTPException(status_code=status_code, detail=error_response_body(e))


# ── Emergency pause (admin only) ─────────────────────────


class PauseRequest(BaseModel):
    paused: Optional[bool] = True


@app.get("/api/whitelist")
async def get_whitelist():
    """Return the current protocol whitelist."""
    return {
        "whitelist": list(PROTOCOL_WHITELIST),
        "mode": "permissive" if not PROTOCOL_WHITELIST else "enforced",
    }


class WhitelistRequest(BaseModel):
    protocol_address: str
    action: str = "add"  # "add" or "remove"


@app.post("/api/whitelist")
async def manage_whitelist(req: WhitelistRequest):
    """Add or remove a protocol from the backend whitelist."""
    if not ADMIN_CAP_ID:
        raise HTTPException(status_code=403, detail="Admin key not configured")
    if req.action == "add":
        PROTOCOL_WHITELIST.add(req.protocol_address)
        logger.info(f" Added {req.protocol_address} to protocol whitelist")
    elif req.action == "remove":
        PROTOCOL_WHITELIST.discard(req.protocol_address)
        logger.info(f" Removed {req.protocol_address} from protocol whitelist")
    else:
        raise HTTPException(status_code=400, detail="action must be 'add' or 'remove'")
    return {"whitelist": list(PROTOCOL_WHITELIST), "action": req.action}


@app.post("/api/pause")
async def pause(req: PauseRequest):
    if not ADMIN_CAP_ID:
        raise HTTPException(status_code=403, detail="Admin key not configured on this relayer")

    try:
        cmd = build_set_paused(req.paused)
        result = _run_sui_cmd(cmd)

        if not result.success:
            raise HTTPException(status_code=500, detail=error_response_body(result.error))

        label = "PAUSED" if req.paused else "RESUMED"
        logger.info(f"{label} TX: {result.digest}")

        return {
            "success": True,
            "digest": result.digest,
            "paused": req.paused,
        }

    except HTTPException:
        raise
    except Exception as e:
        log_error("Pause", e)
        raise HTTPException(status_code=500, detail=error_response_body(e))


# ═══════════════════════════════════════════════════════════
#  START
# ═══════════════════════════════════════════════════════════

if __name__ == "__main__":
    import uvicorn

    print(f"\n Relayer server running on http://localhost:{PORT}")
    print(f"   Network:    {NETWORK}")
    print(f"   Package:    {PACKAGE_ID}")
    print(f"   Portfolio:  {PORTFOLIO_ID}")
    print(f"\n   Endpoints:")
    print(f"     POST /api/trade             — Execute quantum-optimized trade")
    print(f"     POST /api/atomic-rebalance  — Multi-swap atomic rebalance")
    print(f"     POST /api/oracle-swap       — Oracle-validated swap")
    print(f"     POST /api/audit             — Log quantum proof on-chain")
    print(f"     POST /api/pause             — Emergency kill-switch")
    print(f"     GET  /api/status            — Portfolio & gas status")
    print(f"     GET  /api/health            — Health check\n")

    uvicorn.run(app, host="0.0.0.0", port=PORT)
